src/pip/_internal/index.py

```diff
@@ -815,18 +815,16 @@ def egg_info_matches(
     if not match:
         logger.debug('Could not parse version from link: %s', link)
         return None
+    full_match = match.group(0)
     if search_name is None:
-        full_match = match.group(0)
         return full_match.split('-', 1)[-1]
-    name = match.group(0).lower()
-    # To match the "safe" name that pkg_resources creates:
-    name = name.replace('_', '-')
-    # project name and version must be separated by a dash
-    look_for = search_name.lower() + "-"
-    if name.startswith(look_for):
-        return match.group(0)[len(look_for):]
-    else:
+    canonicalized_name = canonicalize_name(search_name)
+    if not canonicalize_name(full_match).startswith(canonicalized_name):
+        return None
+    # Project name and version must be separated by a dash.
+    if full_match[len(search_name)] != "-":
         return None
+    return full_match[(len(search_name) + 1):]
```

Comment:
Is there a case to consider if the original match didn't occur at the beginning of the string? For example, it seems like a test case like `("pip-pip-18.0", "pip", "18.0")` would yield version `pip-18.0`.

Reply:
PyPI historically allow packages to specify arbitrary strings as version, so if a package named `pip` has a version `pip-18.0`, `pip-pip-18.0` is exactly the sdist name to expect.

Links on a simple API page are guaranteed to come from a specific name (e.g. `pip` if the URL is `/simple/pip/`, or `pip-pip` if `/simple/pip-pip/`), so the name *should* match the beginning. Ambiguities occur for flat HTML pages (find-links values), but this has been the case since as far as I can recall, and there’s really no way to work around it either.

Reply:
I guess what I meant to ask is if there are any real-world cases where `_egg_info_re.search(egg_info)` doesn't match the full provided `egg_info` string (e.g. starts after the beginning and/or ends before getting to the end) and whether it would make sense to add such test cases, because it looks like all of the test cases match the full string. (The example I provided wasn't such a case after all.)

Reply:
Hmm, I never thought about that. Python package names (as defined in PEP 508) must return a match to

```python
re.match(r'^([A-Z0-9]|[A-Z0-9][A-Z0-9._-]*[A-Z0-9])$', re.IGNORECASE)
```

Currently `_egg_info_re` is `r'([a-z0-9_.]+)-([a-z0-9_.!+-]+)'`, so I think it will always match the whole string for any validly-named links.